### ISSUE TITLE
Texture Streaming 

### ISSUE THREAD

**Issue by OldSnake22** [urho3d/Urho3D#1898]
This is a feature request. Texture streaming would be really nice to have in the engine. Building quite large open world with Urho is difficult in terms of GPU memory consumption because of the lack of texture streaming feature in the engine.

How difficult it would be to implement it ? 

Is there some rather easy solution to lower the GPU memory consumption with the actual engine code ? 

Thanks !

**Comment by cadaver** [comment]
Urho doesn't do the (all) work for you, because it cannot know how a large world is going to be structured and accessed. Obviously if you just instantiate everything naively you could run out of memory.

You can request resources for background loading. File access and png / jpg decode can happen in a background thread, however actual GPU upload must happen in main thread after the load/decode is finished. For dds textures there isn't that much to do before the upload, since it's just a data dump.

You probably need to instantiate pieces of the world as you move around, and discard scene nodes and resources you no longer need.

If someone was interested in writing a general solution for handling this it could be made a part of the Urho engine itself, however I suspect it would lock you down to how that someone imagined an open world game should work, and might not be a good fit for all cases.

**Comment by SprueKit** [comment]
I had luck with using separate Scenes for the LoadAsync family of functions and then transferring their contents to the main scene.

Where I ran into trouble was with releasing resources without mangling background loading. I was liberally loading the tiles in resources-only mode at the slightest hint of need and then only doing the final scene-only load when they were really needed. 

I never arrived at any sane solution before scrapping it. 

The closest to sane I got was:

- A "TileTracker" derived from Object that just knew about the async Scene, the "root" node that came from that scene, and had a list of SharedPtr<Resource> to guard them against being released
- Attaching an extra "scene-source" to the background loader so the event message could send that
    - The TileTracker for the tile received that message and kept a copy of the resource
    - since that was right before the addition of the resource to the cache it kept the resource safe from release until the TileTracker was destroyed or dumped it's resources
- Any calls to ResourceCache::BackgroundLoadResource that already had a loaded resource sent an event that the resource was already there so the TileTracker could record a copy of the resource for safe keeping
- I never looked at what budgets would mean, and I really didn't want to know
- At runtime I cleared 1 resource group of all unused resources once each second, moving onto the next for the next second

It was all pretty nasty. The async loading in Urho worked tons better than I had expected, but Resources weren't quite there.

**Comment by JSandusky** [comment]
I had luck with using separate Scenes for the LoadAsync family of functions and then transferring their contents to the main scene.

Where I ran into trouble was with releasing resources without mangling background loading. I was liberally loading the tiles in resources-only mode at the slightest hint of need and then only doing the final scene-only load when they were really needed.

I never arrived at any sane solution before scrapping it.

The closest to sane I got was:

- A "TileTracker" derived from Object that just knew about the async Scene, the "root" node that came from that scene, and had a list of SharedPtr to guard them against being released
- Attaching an extra "scene-source" to the background loader so the event message could send that
- The TileTracker for the tile received that message and kept a copy of the resource
    - since that was right before the addition of the resource to the cache it kept the resource safe from release until the TileTracker was destroyed or dumped it's resources
- Any calls to ResourceCache::BackgroundLoadResource that already had a loaded resource sent an event that the resource was already there so the TileTracker could record a copy of the resource for safe keeping
    - I never looked at what budgets would mean, and I really didn't want to know
- At runtime I cleared 1 resource group of all unused resources once each second, moving onto the next for the next second

It was all pretty nasty. The async loading in Urho worked tons better than I had expected, but Resources weren't quite there.
